src/librustc_typeck/check/pat.rs

```diff
@@ -226,6 +224,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     _ => true,
                 }
             }
+            // FIXME(or_patterns; Centril | dlrobertson): To keep things compiling
+            // for or-patterns at the top level, we need to make `p_0 | ... | p_n`
+            // a "non reference pattern". For example the following currently compiles:
+            // ```
+            // match &1 {
+            //     e @ &(1...2) | e @ &(3...4) => {}
+            //     _ => {}
+            // }
+            // ```
+            //
+            // We should consider whether we should do something special in nested or-patterns.
```

Comment:
Does this mean things are broken with the current set of changes?

Reply:
I'm not sure. It might possibly mean that there's a language design choice. At least, if there is, then consistency (and compiler simplicity) compels us to use the behavior encoded in this PR as opposed to diverging between the top / nested positions.

If I don't do this change then things break in the `libproc_macro` crate wrt. exhaustiveness checking.

I've mostly left the FIXME for now so I can come back to it once we have the exhaustiveness checking and MIR building in place and the or-patterns feature can be used for real. Evaluating the FIXME is probably easier at that stage. IOW, I think this can be filed under "polish".

Reply:
Resolved in https://github.com/rust-lang/rust/pull/68856 by declaring that the implementation here was correct.